### PR TITLE
Delete webkit prefixes in default CSS

### PR DESCRIPTION
**Pull request by flaviut**
See tmpvar/jsdom#1560.

In particular, this helps because it takes out all the selectors that would cause exceptions to be thrown: webkit prefixes in selectors and case-insensitive attribute selectors.

I've tested this patch by running the test that runs slowly with breakpoints set on "all exceptions". There were no exceptions arising from <https://github.com/dperini/nwmatcher/blob/29a1cdc284e330ec6aa38e31c2f02e755a536381/src/nwmatcher-noqsa.js#L765>.

Even though this deletes code, this adds functionality. For example, the rule `-input[type="hidden" i]` would have been ignored before, because nwmatcher doesn't know how to handle ` i`. It would now apply (although type would have to be "hidden", and not "HiDdEn" or "HIDDEN").

This single change improves performance in one of my tests from 28s to 8s.

I'm not really sure how to run the benchmarks. `node ./benchmark/runner` appears to do nothing.

### CODE REVIEW

File: lib/jsdom/browser/default-stylesheet.js [hunk 1]
```diff
@@ -30,8 +30,6 @@ module.exports = `
  *
  */
 
-@namespace "http://www.w3.org/1999/xhtml";
```
Comment (domenic) [inline lib/jsdom/browser/default-stylesheet.js]: This should be kept per spec; did it cause problems with our implementation?

File: lib/jsdom/browser/default-stylesheet.js [hunk 2]
```diff
@@ -128,14 +123,6 @@ q {
     display: inline
 }
 
-q:before {
-    content: open-quote;
```
Comment (domenic) [inline lib/jsdom/browser/default-stylesheet.js]: Why are these removed?

File: lib/jsdom/browser/default-stylesheet.js [hunk 3]
```diff
@@ -173,36 +160,12 @@ h1 {
...
 }
 
-:-webkit-any(article,aside,nav,section) h1 {
+article h1, aside h1, nav h1, section h1 {
     font-size: 1.5em;
     -webkit-margin-before: 0.83__qem;
     -webkit-margin-after: 0.83em;
 }
 
-:-webkit-any(article,aside,nav,section) :-webkit-any(article,aside,nav,section) h1 {
```
Comment (domenic) [inline lib/jsdom/browser/default-stylesheet.js]: Will nwmatcher work with `:matches()` instead of `:-webkit-any()`?

If not, we should add a comment here explaining that we are going against spec for now.

File: lib/jsdom/browser/default-stylesheet.js [hunk 4]
```diff
@@ -423,7 +386,7 @@ input, textarea, keygen, select, button {
     text-align: start;
 }
 
-input[type="hidden" i] {
+input[type="hidden"] {
```
Comment (domenic) [inline lib/jsdom/browser/default-stylesheet.js]: We need to add a TODO comment for these cases where we go against spec

File: lib/jsdom/browser/default-stylesheet.js [hunk 5]
```diff
@@ -437,95 +400,15 @@ input {
...
-    opacity: 0;
-    pointer-events: none;
-}
-
-input:enabled:read-write:-webkit-any(:focus,:hover)::-webkit-inner-spin-button {
-    opacity: 1;
-    pointer-events: auto;
-}
-
 keygen, select {
```
Comment (domenic) [inline lib/jsdom/browser/default-stylesheet.js]: Maybe just remove keygen while you're here

**Comment by dperini** [comment]
Could it be that muting the exception helps in these cases so that execution could continue ?
By default 'nwmatcher' is configured to throw exceptions for CSS syntax errors, however this behaviour can be changed both before execution and at runtime by issuing a:

    NW.Dom.configure({ VERBOSITY: false });

I recently suggested the same for a different parsing problem with similar requirements.
Hope i did correctly understand the OP and the problem posed above.

I will also have an attempt to implement the ' i ' flag for insensitive attributes matching.

**Comment by domenic** [comment]
@dperini, that's a great point. When parsing CSS stylesheets we should definitely mute such errors. I think we currently try/catch them, causing the slowness @flaviut refers to in his OP.

File: lib/jsdom/browser/default-stylesheet.js (hunk 2 above)
Comment (flaviut) [inline lib/jsdom/browser/default-stylesheet.js]: nwmatcher [does not support `:before` or `:after`](https://github.com/dperini/nwmatcher/wiki/CSS-supported-selectors#content-pseudo-element-selectors-not-supported)

File: lib/jsdom/browser/default-stylesheet.js (hunk 3 above)
Comment (flaviut) [inline lib/jsdom/browser/default-stylesheet.js]: Nope, `:matches()` is CSS4, and [nwmatcher only supports CSS3](https://github.com/dperini/nwmatcher/wiki/CSS-supported-selectors#content-pseudo-element-selectors-not-supported):

>The list has been built upon current W3C Selectors Level 3, but differs in that it contains some deprecated specification that we needed to temporarily maintain.

Going through the source code confirms that `:matches()` is not implemented.

> If not, we should add a comment here explaining that we are going against spec for now.

Done!

File: lib/jsdom/browser/default-stylesheet.js (hunk 1 above)
Comment (flaviut) [inline lib/jsdom/browser/default-stylesheet.js]: It did, see https://gist.github.com/flaviut/9485daa073065e5944d769fd6900138f#file-errors-txt-L1

I've restored it though. It seems like a bug that could be fixed in another way.

File: lib/jsdom/browser/default-stylesheet.js (hunk 4 above)
Comment (flaviut) [inline lib/jsdom/browser/default-stylesheet.js]: Done!

File: lib/jsdom/browser/default-stylesheet.js (hunk 5 above)
Comment (flaviut) [inline lib/jsdom/browser/default-stylesheet.js]: I'd rather not -- kegen is still in the spec, and it doesn't break anything.

Comment (domenic) [inline lib/jsdom/browser/default-stylesheet.js]: keygen was removed from the spec in January: https://gist.github.com/domenic/ec8b0fc8ab45f39403dd

**Comment by flaviut** [comment]
Thanks for the quick feedback, I really appreciate it!

`NW.Dom.configure({ VERBOSITY: false });` won't work for me, unless there is some way to mute the messages. It prints a crazy amount of stuff to the console:

```console
$ jest MyTest > out
<snip>
$ wc -l out
131136 out
```

File: lib/jsdom/browser/default-stylesheet.js (hunk 5 above)
Comment (flaviut) [inline lib/jsdom/browser/default-stylesheet.js]: Ok then, removed.

**Comment by dperini** [comment]
@flaviut what if I would add an extra configuration flag to also mute console errors ?
Would this fix your issue ?

I am thinking to name the flag `LOGERRORS` and assign `true` to it as the default value to maintain backward compatibility. If this helps, the change would be trivial and could happen immediately.
/cc @domenic

**Comment by flaviut** [comment]
@dperini It absolutely would solve my issue. Actually, that what I think I'm going to do for now as a temporary workaround--deleting those three lines in node_modules is much easier that applying this patch.

File: lib/jsdom/browser/default-stylesheet.js [hunk 6]
```diff
@@ -128,13 +124,10 @@ q {
...
 
-q:before {
-    content: open-quote;
-}
-
-q:after {
-    content: close-quote;
-}
+/* nwmatcher does not support :before and :after, so we can't render q
+correctly: https://www.w3.org/TR/html5/rendering.html#phrasing-content-0
```
Comment (domenic) [inline lib/jsdom/browser/default-stylesheet.js]: Wrong spec; you want https://html.spec.whatwg.org/multipage/ not the W3C fork. I can fix while merging.

**Comment by dperini** [comment]
@flaviut 
the requested change was released with version 1.4.0 (the. flag name is LOGERRORS).
Setting this flag to `false` will mute all the errors that were previously logged to the console.

File: lib/jsdom/browser/default-stylesheet.js (hunk 2 above)
Comment (dperini) [inline lib/jsdom/browser/default-stylesheet.js]: @domenic 
now nwmatcher does support the :after and :before pseudo-elements and a few others too, the supports is obviously limited to their parsing (grammar) without throwing exceptions.

For these pseudo-elements, the `select()` method will return an empty node set (empty array) while the `match()` method will always return `false`.

Comment (dperini) [reply lib/jsdom/browser/default-stylesheet.js]: @domenic I have added the case insensitive flag for attributes, the `i` flag from the Selectors Level 4 specs have a look at current branch, also the new `nwsapi`will be ready to ship sometime next week with all the bells a whistles, I will also be able to port `:matches` to current branch.
